src/librbd/cache/pwl/Types.h

```diff
@@ -184,31 +200,38 @@ class DeferredContexts {
 };
 
 /* Pmem structures */
+#ifdef WITH_RBD_RWL
 POBJ_LAYOUT_BEGIN(rbd_pwl);
 POBJ_LAYOUT_ROOT(rbd_pwl, struct WriteLogPoolRoot);
 POBJ_LAYOUT_TOID(rbd_pwl, uint8_t);
 POBJ_LAYOUT_TOID(rbd_pwl, struct WriteLogPmemEntry);
 POBJ_LAYOUT_END(rbd_pwl);
+#endif
 
 struct WriteLogPmemEntry {
   uint64_t sync_gen_number = 0;
   uint64_t write_sequence_number = 0;
   uint64_t image_offset_bytes;
   uint64_t write_bytes;
+  #ifdef WITH_RBD_RWL
   TOID(uint8_t) write_data;
+  #endif
+  #ifdef WITH_RBD_SSD_CACHE
+  uint64_t write_data_pos; /* SSD data offset */
+  #endif
   struct {
-    uint8_t entry_valid :1; /* if 0, this entry is free */
-    uint8_t sync_point :1;  /* No data. No write sequence number. Marks sync
+    uint8_t entry_valid; /* if 0, this entry is free */
```

Comment:
The RWL cache was packing these into a single byte -- any reason to break these out into 8 bytes?

Reply:
You can use a union w/ a `uint8_t` I would think, right?

Reply:
yeah but that would mean all the members inside union would use the same address and only one of them is active at a time. But in current code, we simultaneously check for presence/absence of two or more of the fields, so that would create an issue

Reply:
Not if you packed the bitfields in a struct and placed the struct in the union, right?

```
union {
  uint8_t flags;
  struct {
    uint8_t entry_valid :1; /* if 0, this entry is free */
    uint8_t sync_point :1;  /* No data. No write sequence number. Marks sync
                               point for this sync gen number */
    uint8_t sequenced :1;   /* write sequence number is valid */
    uint8_t has_data :1;    /* write_data field is valid (else ignore) */
    uint8_t discard :1;     /* has_data will be 0 if this is a discard */
    uint8_t writesame :1;   /* ws_datalen indicates length of data at write_bytes */
  };
};
```

Reply:
Done, updated code